buffer-memory-segment/src/main/java/io/netty5/buffer/memseg/NativeSegmnentAllocator.java

```diff
@@ -0,0 +1,136 @@
+                                               FunctionDescriptor.of(ValueLayout.ADDRESS, ValueLayout.JAVA_LONG));
+
+                final MemorySegment freeAddress = symbolLookup.find("free").orElseThrow(
+                        () -> new NoSuchElementException("free symbol not found"));
+                free = linker.downcallHandle(freeAddress, FunctionDescriptor.ofVoid(ValueLayout.ADDRESS));
+
+                available = true;
+                logger.debug("direct malloc/free: available");
+            } catch (Throwable t) {
+                available = false;
+                if (logger.isTraceEnabled()) {
+                    logger.debug("direct malloc/free: unavailable", t);
+                } else {
+                    logger.debug("direct malloc/free: unavailable: {}", t.getMessage());
+                }
+            }
+
+            AVAILABLE = available;
+            MALLOC = malloc;
+            FREE = free;
+        }
+    }
+
+    /**
+     * Check if native methods are explicitly disabled, either by providing a system property or by
+     * disabling native access for the module.
+     * @return If native segment allocatiuon is explicitly disabled.
+     */
+    private static boolean explicitlyDisabled() {
+        final boolean noMalloc = SystemPropertyUtil.getBoolean("io.netty5.memseg.noMalloc", false);
+        logger.debug("-Dio.netty5.memseg.noMalloc: {}", noMalloc);
+        if (noMalloc) {
+            logger.debug("direct malloc/free: unavailable (-Dio.netty5.memseg.noMalloc)");
+            return true;
+        }
+
+        final Module module = NativeSegmnentAllocator.class.getModule();
+        final boolean nativeAccessEnabled = module.isNativeAccessEnabled();
+        logger.debug("--enable-native-access for {}: {}", module, nativeAccessEnabled);
+        if (!nativeAccessEnabled) {
+            logger.debug("direct malloc/free: unavailable (native access not enabled for {})", module);
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Returns {@code true} if this allocation method is available.
+     */
+    public static boolean isAvailable() {
+        return AVAILABLE;
+    }
+
+    @Override
+    public SegmentHolder allocate(long byteSize) {
+        final MemorySegment segment;
+        try {
+            segment = (MemorySegment) MALLOC.invokeExact(byteSize);
+        } catch (Throwable throwable) {
+            throw new Error(throwable); // Should not happen
+        }
+        if (segment.address() == 0L) {
+            throw new OutOfMemoryError("malloc(2) failed to allocate " + byteSize + " bytes");
+        }
+        final MemorySegment scaledSegment = segment.reinterpret(byteSize);
+        return new NativeSegmentHolder(scaledSegment);
+    }
+
+    private record NativeSegmentHolder(MemorySegment segment) implements SegmentHolder {
+        @Override
+        public void free() {
+            try {
+                FREE.invokeExact(segment);
```

Comment:
should we maybe add an additional check here to avoid double free, in case upstream code is buggy? Also in the 4.2 PR

Reply:
This should be guaranteed by the buffer drop (in this case an [ArcDrop](https://github.com/netty/netty/blob/main/buffer/src/main/java/io/netty5/buffer/internal/ArcDrop.java#L67)).

Reply:
I don't think this is necessary. We have the shared Arena implementation guarding for this and giving confidence to our ArcDrop, and below you can often configure system allocators to do sanity checks as well, that will catch some cases.